Model/CustomerPaymentMethodsManagement.php

```diff
@@ -0,0 +1,95 @@
+use Magento\Vault\Api\PaymentTokenManagementInterface;
+use Paytrail\PaymentService\Api\CustomerPaymentMethodsManagementInterface;
+use Psr\Log\LoggerInterface;
+
+class CustomerPaymentMethodsManagement implements CustomerPaymentMethodsManagementInterface
+{
+    /**
+     * @var Session
+     */
+    protected $customerSession;
+
+    /**
+     * @var PaymentTokenManagementInterface
+     */
+    protected $paymentTokenManagement;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    protected $searchCriteriaBuilder;
+
+    /**
+     * @var Json
+     */
+    protected $jsonSerializer;
+
+    /**
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * @param Session $customerSession
+     * @param PaymentTokenManagementInterface $paymentTokenManagement
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param Json $jsonSerializer
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        Session                         $customerSession,
+        PaymentTokenManagementInterface $paymentTokenManagement,
+        SearchCriteriaBuilder           $searchCriteriaBuilder,
+        Json                            $jsonSerializer,
+        LoggerInterface                 $logger
+    ) {
+        $this->customerSession = $customerSession;
+        $this->paymentTokenManagement = $paymentTokenManagement;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->jsonSerializer = $jsonSerializer;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @return \Paytrail\PaymentService\Api\Data\CustomerPaymentMethodsResultInterface[]
+     * @throws LocalizedException
+     */
+    public function showCustomerPaymentMethods(): array
+    {
+        $paymentMethods = [];
+        try {
+            if ($this->customerSession->isLoggedIn()) {
+                $customerId = $this->customerSession->getCustomerId();
+                $tokens = $this->paymentTokenManagement->getListByCustomerId($customerId);
+
+                foreach ($tokens as $token) {
+                    if ($token->getIsActive() && $token->getIsVisible()) {
+                        $paymentMethods[] = [
+                            'entity_id' => $token->getId(),
+                            'customer_id' => $customerId,
+                            'type' => $token->getType(),
+                            'payment_method_code' => $token->getPaymentMethodCode(),
+                            'created_at' => $token->getCreatedAt(),
+                            'expires_at' => $token->getExpiresAt(),
+                            'card_type' => $this->jsonSerializer->unserialize($token->getTokenDetails())['type'],
+                            'maskedCC' => $this->jsonSerializer->unserialize($token->getTokenDetails())['maskedCC']
+                        ];
+                    }
+                }
+            }
+
+            return $paymentMethods;
```

Comment:
Returning a multidimensional array instead of the expect array of \Paytrail\PaymentService\Api\Data\CustomerPaymentMethodsResultInterface objects.
Please use a factory to create the necessary objects and add the data into each object instead.
```$token = $this->CustomerMethodsResultsFactory->create();
$token->setEntityId($token->getId())
    ->setCustomerId($customer)
    ->set...
$paymentMethods[] = $token;
```

Reply:
Done